{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tasks.scfoundation import load\n",
    "import scipy.sparse as sp\n",
    "import anndata as ad\n",
    "import hdf5plugin\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc \n",
    "from scipy.spatial.distance import cdist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_data(adata, need_normalization=True):\n",
    "    gene_list_df = pd.read_csv(f'{tokenizer_dir}/OS_scRNA_gene_index.19264.tsv', header=0, delimiter='\\t')\n",
    "    gene_list = list(gene_list_df['gene_name'])\n",
    "    adata = adata[:,adata.var['gene_name'].isin(gene_list)]\n",
    "    \n",
    "    gexpr_feature = adata.X.A\n",
    "    idx = adata.obs_names.tolist()\n",
    "    col = adata.var['gene_name'].tolist()\n",
    "    gexpr_feature = pd.DataFrame(gexpr_feature, index=idx, columns=col)\n",
    "    gexpr_feature, _ = load.main_gene_selection(gexpr_feature, gene_list)\n",
    "\n",
    "    varnames = gexpr_feature.columns.values\n",
    "    matrix = sp.csr_matrix(gexpr_feature.values)\n",
    "    adata_pp = ad.AnnData(X=matrix, obs=adata.obs, var=pd.DataFrame({}, index=varnames), obsm=adata.obsm)\n",
    "\n",
    "    if need_normalization:\n",
    "        sc.pp.normalize_total(adata_pp, target_sum=1e4)\n",
    "        sc.pp.log1p(adata_pp)\n",
    "\n",
    "    return(adata_pp)\n",
    "\n",
    "def get_niche_samples_sc(adata, ligand_index):\n",
    "    adata.uns['cell_types_list'] = adata.obs['cell_type'].cat.categories.tolist()\n",
    "    cell_types_index = adata.obs['cell_type'].cat.codes.values+1\n",
    "    adata.obs['cell_type'] = cell_types_index\n",
    "\n",
    "    all_counts = adata.X.A\n",
    "    niche_composition = np.zeros([adata.shape[0], 50])\n",
    "    niche_celltypes = np.zeros([adata.shape[0], 50])\n",
    "    niche_ligands_expression = []\n",
    "\n",
    "    for i in range(adata.shape[0]):\n",
    "        niche_composition[i, 0] = 1\n",
    "\n",
    "        niche_celltypes[i, 0] = cell_types_index[i]\n",
    "\n",
    "        ligands_expression = np.zeros([50, len(ligand_index)])\n",
    "        ligands_expression[0] = all_counts[i, ligand_index]\n",
    "        ligands_expression = np.concatenate(ligands_expression).reshape(1,-1)\n",
    "        niche_ligands_expression.append(ligands_expression)\n",
    "\n",
    "    niche_composition = sp.csr_matrix(niche_composition)\n",
    "    niche_ligands_expression = sp.csr_matrix(np.concatenate(niche_ligands_expression))\n",
    "    niche_celltypes = sp.csr_matrix(niche_celltypes)\n",
    "\n",
    "    adata.X = adata.X.astype(np.float32)\n",
    "    adata.obsm['niche_composition'] = niche_composition.astype(np.float32)\n",
    "    adata.obsm['niche_ligands_expression'] = niche_ligands_expression.astype(np.float32)\n",
    "    adata.obsm['niche_celltypes'] = niche_celltypes.astype(np.float32)\n",
    "    \n",
    "    return(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_niche_radius(spatial_coords, max_niche_cell_num):\n",
    "    distances = cdist(spatial_coords, spatial_coords)\n",
    "    radius = 0\n",
    "    counts = np.sum(distances < radius, axis=1)\n",
    "    while max(counts) <= max_niche_cell_num:\n",
    "        radius += 0.1\n",
    "        counts = np.sum(distances < radius, axis=1)\n",
    "    return radius-0.1\n",
    "\n",
    "def get_niche_samples(adata, spatial_coords, niche_radius, ligand_index, max_niche_cell_num=None):\n",
    "    all_counts = adata.X.A\n",
    "    cell_types_index = adata.obs['cell_type'].values\n",
    "\n",
    "    niche_ligands_expression = []\n",
    "    niche_composition = []\n",
    "    niche_celltypes = np.zeros([adata.shape[0], 50])\n",
    "\n",
    "    distances = cdist(spatial_coords, spatial_coords)\n",
    "    for i in range(len(all_counts)):\n",
    "        distances_i = distances[i]\n",
    "\n",
    "        if niche_radius is None:\n",
    "            niche_cells = np.argsort(distances_i)[:max_niche_cell_num]\n",
    "        else:\n",
    "            niche_cells = np.where(distances_i <= niche_radius)[0]\n",
    "        \n",
    "        niche_counts = np.concatenate([all_counts[niche_cells][:,ligand_index], np.zeros([50-len(niche_cells), 986])])\n",
    "        niche_counts = np.concatenate(niche_counts)\n",
    "        niche_ligands_expression.append(np.expand_dims(niche_counts, axis=0))\n",
    "\n",
    "        niche_prop = np.concatenate([np.array([1]*len(niche_cells)), np.array([0]*(50-len(niche_cells)))])\n",
    "        niche_prop = niche_prop/niche_prop.sum()\n",
    "        niche_composition.append(np.expand_dims(niche_prop, axis=0))\n",
    "\n",
    "        niche_celltypes[i, :len(niche_cells)] = cell_types_index[niche_cells]\n",
    "\n",
    "    niche_ligands_expression = sp.csr_matrix(np.concatenate(niche_ligands_expression))\n",
    "    niche_composition = sp.csr_matrix(np.concatenate(niche_composition))\n",
    "    niche_celltypes = sp.csr_matrix(niche_celltypes)\n",
    "\n",
    "    adata.X = adata.X.astype(np.float32)\n",
    "\n",
    "    if max_niche_cell_num is None:\n",
    "        adata.obsm['niche_composition'] = niche_composition.astype(np.float32)\n",
    "        adata.obsm['niche_ligands_expression'] = niche_ligands_expression.astype(np.float32)\n",
    "        adata.obsm['niche_celltypes'] = niche_celltypes.astype(np.float32)\n",
    "    else:\n",
    "        adata.obsm[f'niche_composition_niche{max_niche_cell_num}'] = niche_composition.astype(np.float32)\n",
    "        adata.obsm[f'niche_ligands_expression_niche{max_niche_cell_num}'] = niche_ligands_expression.astype(np.float32)\n",
    "        adata.obsm[f'niche_celltypes_niche{max_niche_cell_num}'] = niche_celltypes.astype(np.float32)\n",
    "    \n",
    "    return(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer_dir = '../stformer/tokenizer/'\n",
    "\n",
    "gene_list_df = pd.read_csv(f'{tokenizer_dir}/OS_scRNA_gene_index.19264.tsv', header=0, delimiter='\\t')\n",
    "ligand_database = pd.read_csv(tokenizer_dir+'ligand_database.csv', header=0, index_col=0)\n",
    "ligand_symbol = ligand_database[ligand_database.sum(1)>1].index.values\n",
    "ligand_index = gene_list_df.loc[gene_list_df['gene_name'].isin(ligand_symbol), 'index'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = 'pancreas_cosmx_rds'\n",
    "\n",
    "adata = sc.read_mtx(f'{dataset}/count_matrix.mtx').T\n",
    "\n",
    "df1 = pd.read_csv(f'{dataset}/genes.tsv', sep='\\t', header=None)\n",
    "df2 = pd.read_csv(f'{dataset}/barcodes.tsv', sep='\\t', header=None)\n",
    "df3 = pd.read_csv(f'{dataset}/meta_data.csv', index_col=0)\n",
    "\n",
    "adata.var_names = df1[0]\n",
    "adata.obs_names = df2[0]\n",
    "adata.var['gene_name'] = df1[0].values\n",
    "adata.obs = df3.loc[adata.obs_names]\n",
    "\n",
    "adata.write(f'{dataset}/pancreas_cosmx.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[adata.obs['types']!='QC_dropped']\n",
    "adata.obs['cell_type'] = adata.obs['types'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = prepare_data(adata, need_normalization=True)\n",
    "\n",
    "# #--------- Prepare sample data from single-cell RNA-seq data ------------------\n",
    "# adata = get_niche_samples_sc(adata, ligand_index)\n",
    "\n",
    "\n",
    "#--------- Prepare sample data from single-cell spatial data ------------------\n",
    "cell_types_list = adata.obs['cell_type'].cat.categories.tolist()\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].cat.codes.values+1\n",
    "\n",
    "for max_niche_cell_num in [5,10,15,20,25,30,35]:\n",
    "    adata_list = []\n",
    "\n",
    "    for fov in set(adata.obs['fov'].value_counts().index.values):\n",
    "        adata0 = adata[adata.obs['fov']==fov]\n",
    "\n",
    "        spatial_coords = adata0.obs[['x_FOV_px', 'y_FOV_px']].values # spatial_coords = adata0.obsm['spatial']\n",
    "        niche_radius = get_niche_radius(spatial_coords, max_niche_cell_num) # niche_radius = None\n",
    "        print(fov, niche_radius)\n",
    "        adata0 = get_niche_samples(adata0, spatial_coords, niche_radius, ligand_index, max_niche_cell_num)\n",
    "\n",
    "        adata_list.append(adata0)\n",
    "\n",
    "    adata = sc.concat(adata_list)\n",
    "\n",
    "adata.uns['cell_types_list'] = cell_types_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(f'pancreas_cosmx_niche.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "stformer",
   "language": "python",
   "name": "stformer"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
